larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewCosmicRayRemovalTool.cc

```diff
@@ -0,0 +1,560 @@
+    
+    const float transverseDistanceFromLine(lineDirection.GetCrossProduct(hitPosition - lineStart).GetMagnitude());
+    
+    if (transverseDistanceFromLine > distanceToLine)
+       return false;
+
+    return true;
+}    
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewCosmicRayRemovalTool::IsInLineSegment(const CartesianVector &lowerBoundary, const CartesianVector &upperBoundary, const CartesianVector &point) const
+{
+    const float segmentBoundaryGradient = (-1.f) * (upperBoundary.GetX() - lowerBoundary.GetX()) / (upperBoundary.GetZ() - lowerBoundary.GetZ());
+    const float xPointOnUpperLine((point.GetZ() - upperBoundary.GetZ()) / segmentBoundaryGradient + upperBoundary.GetX());
+    const float xPointOnLowerLine((point.GetZ() - lowerBoundary.GetZ()) / segmentBoundaryGradient + lowerBoundary.GetX());
+    
+    if (std::fabs(xPointOnUpperLine - point.GetX()) < std::numeric_limits<float>::epsilon())
+        return true;
+
+    if (std::fabs(xPointOnLowerLine - point.GetX()) < std::numeric_limits<float>::epsilon())
+        return true;
+    
+    if ((point.GetX() > xPointOnUpperLine) && (point.GetX() > xPointOnLowerLine))
+        return false;
+
+    if ((point.GetX() < xPointOnUpperLine) && (point.GetX() < xPointOnLowerLine))
+        return false;
+
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewCosmicRayRemovalTool::IsBestElement(const MatrixType::Element &element, 
+    const HitType hitType, const MatrixType::ElementList &elementList) const
+{
+    const  float chiSquared(element.GetOverlapResult().GetReducedChiSquared());
+    const unsigned int hitSum(element.GetCluster1()->GetNCaloHits() + element.GetCluster2()->GetNCaloHits());
+            
+    for (const MatrixType::Element &testElement : elementList)
+    {
+        if (m_pParentAlgorithm->GetCluster(testElement, hitType) != m_pParentAlgorithm->GetCluster(element, hitType))
+            continue;
+
+        if ((testElement.GetCluster1() == element.GetCluster1()) && (testElement.GetCluster2() == element.GetCluster2()))
+            continue;
+
+        const  float testChiSquared(testElement.GetOverlapResult().GetReducedChiSquared());
+        const unsigned int testHitSum(testElement.GetCluster1()->GetNCaloHits() + testElement.GetCluster2()->GetNCaloHits());
+
+        if ((testHitSum < hitSum) || ((testHitSum == hitSum) && (testChiSquared > chiSquared)))
+            continue;
+
+        if (this->PassElementChecks(testElement, hitType))
+            return false;
+    }
+
+    return true;
+}    
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TwoViewCosmicRayRemovalTool::CreateSeed(const MatrixType::Element &element,
```

Comment:
Very similar to CosmicRayRemovalTool. Again appears to be the difference between TensorType and MatrixType that is causing this duplication. Once you've had a chance to confirm if this is driving these differences, we can potentially try to take a look at getting these types to offer a (relatively) common public interface as a way to eliminate this.